### PR TITLE
⏲ Re-engineer path tracking and proxification

### PR DESCRIPTION
**Pull request by alshakero**
The issue (https://github.com/Palindrom/JSONPatcherProxy/issues/16) arose with modifying a big array. JSONPatcherProxy traversed the whole tree N^3 times for any mutation that changes the array's length! For an array of N=100, this is 1000000 times of intensive iterations. 

I re-designed the whole path tracking method and now JSONPatcherProxy is 65x (6591%) faster when it comes to that issue.

Proxification recursion was very bad and was actually wrong, now it's much cleaner and 586% faster than before. 

Modifying a realistic (size-wise) object (less than 50 children) is much faster than jsonpatch. It's 449% faster than jsonpatch and 281% faster than JSONPatcherProxy before.

Here is a chart for everything:
image

Numbers before: https://travis-ci.org/Palindrom/JSONPatcherProxy/builds/272802078
Numbers after: https://travis-ci.org/Palindrom/JSONPatcherProxy/builds/272793569

## How this works:

It's much simpler than before. Now whenever an object is proxified I record its
- `parent`
- `path`

Note that the `parent` was also proxified at one point of time, so we know its path, and so on.

### Now, when an object is mutated, in a way that doesn't change its hierarchy (`arr[1].name = 1`), we:
- Go to our (`obj => {parent, path}`) map.
- Find the object itself, collect its `path` (basically its name, as in its `key` within its parent object). Collect its `parent` object, too.
- Now that we have the `parent`, we look up its `name` too. And we do this until we arrive at `root`. We `join` these keys with `/` delimiter.

This is very cheap because it happens DEPTH times. Independent of array size. A mega elements wouldn't matter.

### When a change that changes hierarchy occurs:

This happens when we remove say the first element of a big array. All other elements will need their paths updated. (`arr[1]` is now `arr[0]`). In this case, the complexity is O(N), and there is no sync way out of it. `array.shift()` doesn't remove the first item, it literally shifts the whole array. When the array is being shifted, element by element (this has nothing to do with our impl) we update each child's name (`key`) to the new one with an O(1) operation. O(N) * O(1) = O(N).

### CODE REVIEW

**Comment by tomalec** [comment]
Could you describe briefly how the new flow works now, as diff is quite big? :)

**Comment by alshakero** [comment]
Sure! I updated the original comment.

File: src/jsonpatcherproxy.js [hunk 1]
```diff
@@ -150,176 +84,202 @@ const JSONPatcherProxy = (function() {
...
+    this.isObserving = true;
+  }
+  function pause() {
+    this.defaultCallback = () => {};
+    this.isObserving = false;
+  }
+  /**
+    * Creates an instance of JSONPatcherProxy around your object of interest `root`. 
+    * @param {Object|Array} root - the object you want to wrap
+    * @param {Boolean} [showDetachedWarning] - whether to log a warning when a detached sub-object is modified @see {@link https://github.com/Palindrom/JSONPatcherProxy#detached-objects}
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: is `true` a default value? if so I'd use the `[showDetachedWarning=true]` syntax

File: test/spec/proxyBenchmark.js [hunk 2]
```diff
@@ -8,11 +8,11 @@ if (typeof window === 'undefined') {
 }
 
 if (typeof jsonpatch === 'undefined') {
-  var jsonpatch = require('fast-json-patch');
```
Comment (tomalec) [inline test/spec/proxyBenchmark.js]: is there any reason why you removed this `var`?

File: test/spec/proxySpec.js [hunk 3]
```diff
@@ -1146,56 +1142,7 @@ describe('proxy', function() {
...
-
-      //change the array in second observer
-      observedObjTwo.three[2] = 100;
-
-      // expect a patch from second observer (replace /three/2/)
-      expect(countTwo).toReallyEqual(2);
-
-      // first observer should NOT emit a patch
-      expect(countOne).toReallyEqual(2);      
-    });
```
Comment (tomalec) [inline test/spec/proxySpec.js]: Have we changed the behavior? are we proxify already proxified objects?

File: test/spec/proxyBenchmark.js (hunk 2 above)
Comment (alshakero) [reply test/spec/proxyBenchmark.js]: Yes, running tests in browser causes a collision because `jsonpatch` is defined and `var jsonpatch` is hoisted. And btw I added `var` and then removed it. So it wasn't there for a reason.

File: test/spec/proxySpec.js (hunk 3 above)
Comment (alshakero) [reply test/spec/proxySpec.js]: Yup. I removed deep cloning from here and from initial proxification for better performance.

File: src/jsonpatcherproxy.js [hunk 4]
```diff
@@ -150,176 +84,202 @@ const JSONPatcherProxy = (function() {
...
+      typeof newValue == 'object' &&
+      !instance.proxifiedObjectsMap.has(newValue)
+    ) {
+      instance.parenthoodMap.set(newValue, {
+        parent: target,
+        path: key
+      });
+      newValue = instance._proxifyObjectTreeRecursively(target, newValue, key);
+    }
+    if (typeof newValue == 'undefined') {
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: Don't we need to delete an object from parenthood map and disable traps here?

File: src/jsonpatcherproxy.js [hunk 5]
```diff
@@ -76,70 +20,60 @@ const JSONPatcherProxy = (function() {
...
-        because at one point in the future, paths might change and we will simply update our cache instead of 
-        proxifying again.  */
+    return '';
+  }
 
-        var destinationPropKey =
-          instance.getOrSetObjectPath(target, path) +
-          '/' +
-          JSONPatcherProxy.escapePathComponent(key);
+  function setTrap(instance, target, key, newValue) {
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: JSDoc please ;)

File: src/jsonpatcherproxy.js [hunk 6]
```diff
@@ -150,176 +84,202 @@ const JSONPatcherProxy = (function() {
...
+    } else {
+      instance.defaultCallback({
+        op: 'add',
+        path: destinationPropKey,
+        value: newValue
+      });
+      return Reflect.set(target, key, newValue);
+    }
+  }
+  function deleteTrap(instance, target, key) {
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: JSDoc please

File: test/spec/proxyBenchmark.js (hunk 2 above)
Comment (tomalec) [reply test/spec/proxyBenchmark.js]: ehh... I wish we have ES Modules in IE/Edge and FF w/o the flag. Could you add a small note in code comment about it?

File: test/spec/proxySpec.js (hunk 3 above)
Comment (tomalec) [reply test/spec/proxySpec.js]: How would the new version behave if you cross proxify the object?

**Comment by tomalec** [comment]
Great job! we even slightly improved the original performance of our customer's app! :)

File: test/spec/proxySpec.js (hunk 3 above)
Comment (alshakero) [reply test/spec/proxySpec.js]: Both instances would emit patches at once. As one would expect.

File: src/jsonpatcherproxy.js (hunk 4 above)
Comment (alshakero) [reply src/jsonpatcherproxy.js]: Great find!

File: test/spec/proxySpec.js (hunk 3 above)
Comment (tomalec) [reply test/spec/proxySpec.js]: Cool.

How about edge cases, like:
```js
var obj = {
    one: [1, 2, 3, 4, 5],
    two: {
        raz: [6, 7, 6, 5, 4]
    }
};
var observedObj = new JSONPatcherProxy(obj).observe(true, callback);
var observedObjTwo = new JSONPatcherProxy(observedObj.two).observe(true, callback2);

delete observedObj.two.raz;
```
or
```js
var observedObj = new JSONPatcherProxy(obj).observe(true, callback);
var observedObjTwo = new JSONPatcherProxy(obj.two).observe(true, callback2);

delete observedObj.two.raz;
```

File: src/jsonpatcherproxy.js [hunk 7]
```diff
@@ -76,70 +20,66 @@ const JSONPatcherProxy = (function() {
...
-        return Reflect.get(target, propKey, newValue);
-      },
-      set: function(target, key, newValue) {
-        /* each proxified object has its path cached, we need to use that instead of `path` variable
-        because at one point in the future, paths might change and we will simply update our cache instead of 
-        proxifying again.  */
+    return '';
+  }
+  /**
+   * The proxy setter trap, this function is used for all proxies and is pre-defined outside for better performance.
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: I'd rather describe what it does, like 
> update parenthood map if new, proxifies nested objects, calls default callbacks, 

etc...

File: src/jsonpatcherproxy.js [hunk 8]
```diff
@@ -150,176 +90,214 @@ const JSONPatcherProxy = (function() {
...
+        } else {
+          instance.defaultCallback({
+            op: 'remove',
+            path: destinationPropKey
+          });
         }
-        if (Array.isArray(target) && !Number.isInteger(+key.toString())) {
-          return ownReflectSet(instance, target, key, newValue);
+        const oldValue = instance.proxifiedObjectsMap.get(target[key]);
+        // was tje deleted a proxified object?
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: "the"?

File: src/jsonpatcherproxy.js [hunk 9]
```diff
@@ -150,176 +90,214 @@ const JSONPatcherProxy = (function() {
...
-            });
-            return ownReflectSet(instance, target, key, newValue);
-          }
+        return Reflect.set(target, key, newValue);
+      } else if (!Array.isArray(target)) {
+        return Reflect.set(target, key, newValue);
+      }
+    }
+    if (Array.isArray(target) && !Number.isInteger(+key.toString())) {
+      return Reflect.set(target, key, newValue);
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: That means we disregard setting properties to arrays? (if I'm right please add it as a comment, as I'm not 100% sure)

File: src/jsonpatcherproxy.js [hunk 10]
```diff
@@ -150,176 +90,214 @@ const JSONPatcherProxy = (function() {
...
+            path: destinationPropKey
+          });
         }
-        if (Array.isArray(target) && !Number.isInteger(+key.toString())) {
-          return ownReflectSet(instance, target, key, newValue);
+        const oldValue = instance.proxifiedObjectsMap.get(target[key]);
+        // was tje deleted a proxified object?
+        if(oldValue) { 
+          instance.parenthoodMap.delete(target[key]);
+          instance.proxifiedObjectsMap.delete(oldValue);
```
Comment (tomalec) [inline src/jsonpatcherproxy.js]: Don't we have to disable traps for proxy?

File: test/spec/proxySpec.js (hunk 3 above)
Comment (alshakero) [reply test/spec/proxySpec.js]: Good catch. This causes an issue. 

Proxifying an object actually modifies it, because it does
`obj.prop = Proxify(obj.prop)`

So the second constructor would emit patches that are unexpected. Check here. 

https://jsfiddle.net/L0myd8p1/

I'll go back to deep cloning but will find a faster alternative.

File: src/jsonpatcherproxy.js (hunk 9 above)
Comment (alshakero) [reply src/jsonpatcherproxy.js]: You're right. Otherwise, there will be `length` patches all the time.

File: src/jsonpatcherproxy.js (hunk 10 above)
Comment (alshakero) [reply src/jsonpatcherproxy.js]: I thought the GC would take of that. But now I think you're right.

File: test/spec/proxySpec.js (hunk 3 above)
Comment (tomalec) [reply test/spec/proxySpec.js]: maybe we can separate this issue as it's not a natural case. So, we will release fixed performance first.

Comment (alshakero) [reply test/spec/proxySpec.js]: Good point. We shouldn't cater for other people on our own expense. I guess a boolean in the constructor would do.
